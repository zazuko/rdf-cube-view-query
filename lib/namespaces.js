import $rdf from '@zazuko/env'

const cube = $rdf.namespace('https://cube.link/')
const view = $rdf.namespace('https://cube.link/view/')

const {
  rdf,
  schema,
  sh,
  xsd,
} = $rdf.ns

export { cube, rdf, schema, sh, view, xsd }
