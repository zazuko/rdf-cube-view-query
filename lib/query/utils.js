import rdf from '@zazuko/env'

export function contains(ptr, target) {
  if (!target) {
    return ptr.terms.length > 0
  }

  if (target.term) {
    target = target.term
  }

  if (target.ptr && target.ptr.term) {
    target = target.ptr.term
  }

  return ptr.terms.some(term => term.equals(target))
}

export function distinct(...ptrs) {
  const ptr = ptrs[0]
  const termSet = rdf.termSet()

  for (const ptr of ptrs) {
    for (const term of ptr.terms) {
      termSet.add(term)
    }
  }

  const terms = [...termSet]

  if (terms.length === 0) {
    return ptr
  }

  return rdf.clownface({
    dataset: ptr._context[0].dataset,
    term: terms,
    graph: ptr._context[0].graph,
  })
}

export function toPropertyPath(ptr) {
  const list = ptr.list()

  if (!list) {
    return ptr.term
  }

  return [...list].map(property => property.term)
}
