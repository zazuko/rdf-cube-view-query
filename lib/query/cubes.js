import rdf from '@zazuko/env'
import * as sparql from 'rdf-sparql-builder'
import * as ns from '../namespaces.js'

export function cubesQuery({ filters = [], graph = rdf.defaultGraph() } = {}) {
  const cube = rdf.variable('cube')

  const patterns = [[cube, ns.rdf.type, ns.cube.Cube]]

  for (const [index, filter] of filters.entries()) {
    const content = filter({ cube, index })

    for (const c of content) {
      patterns.push(c)
    }
  }

  return sparql.select([cube])
    .from(graph)
    .where(patterns)
    .toString()
}
