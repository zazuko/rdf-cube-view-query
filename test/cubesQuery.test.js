import { deepStrictEqual, strictEqual } from 'assert'
import chai, { expect } from 'chai'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import { cubesQuery } from '../lib/query/cubes.js'
import * as ns from './support/namespaces.js'
import { cleanQuery } from './support/utils.js'

describe('query/cubes', () => {
  chai.use(jestSnapshotPlugin())

  it('should be a function', () => {
    strictEqual(typeof cubesQuery, 'function')
  })

  it('should create a query that finds all cube:Cube terms', () => {
    const query = cubesQuery()

    expect(cleanQuery(query)).toMatchSnapshot()
  })

  it('should use the given named graph', () => {
    const graph = ns.ex.graph

    const query = cubesQuery({ graph })

    expect(cleanQuery(query)).toMatchSnapshot()
  })

  it('should call the filters functions', () => {
    const args = []

    const filter = name => {
      return ({ cube, index }) => {
        args.push({ name, termType: cube.termType, value: cube.value, index })

        return []
      }
    }

    cubesQuery({ filters: [filter('1'), filter('2')] })

    deepStrictEqual(args, [
      { name: '1', termType: 'Variable', value: 'cube', index: 0 },
      { name: '2', termType: 'Variable', value: 'cube', index: 1 },
    ])
  })

  it('should use the result of the filter functions', () => {
    const filter = name => {
      return ({ index }) => {
        return [`# ${name} ${index}`]
      }
    }

    const query = cubesQuery({ filters: [filter('1'), filter('2')] })

    expect(cleanQuery(query)).toMatchSnapshot()
  })
})
