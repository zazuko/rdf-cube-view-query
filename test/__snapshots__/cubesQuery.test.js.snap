// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query/cubes should create a query that finds all cube:Cube terms 1`] = `"SELECT ?cube WHERE { ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>. }"`;

exports[`query/cubes should use the given named graph 1`] = `
"SELECT ?cube FROM <http://example.org/graph>
WHERE { ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>. }"
`;

exports[`query/cubes should use the result of the filter functions 1`] = `"SELECT ?cube WHERE { ?cube <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <https://cube.link/Cube>. }"`;
